### PR TITLE
Live update of stdout/stderr + outputs from running scripts

### PR DESCRIPTION
**Pull request by mfitzp**
Automatically update stdout/stderr and other updates from running jobs.

### CODE REVIEW

**Comment by mfitzp** [comment]
The first commit handles the collection of stderr/stdout and pushing to the database. The next step is implementing the auto-updating of the UI frontend (which is also required to not require a refresh on job completion).

File: wooey/tasks.py [hunk 1]
```diff
@@ -69,9 +94,30 @@ def submit_script(**kwargs):
...
+
+    terr = Thread(target=append_output, args=(proc.stderr, stderr))
+    terr.start()
+
+    prev_std = None
+
+    # Loop until the process is complete + both stdout/stderr have EOFd
+    while proc.poll() is None or tout.is_alive() or terr.is_alive():
+        # Only update if there is output from the script
+        if stdout + stderr != prev_std:
```
Comment (Chris7) [inline wooey/tasks.py]: I think it would be better to use a tuple for this comparison so we remove any possible issues with concatenation

**Comment by Chris7** [comment]
Are you sure we need to use threads for this? I'm fairly sure you can read from stdout/stderr in a manner that doesn't block. I'm always hesitant to rely on python's threading...

Something like this:
```
while True:
        stdout = proc.stdout.readline() # or a chunk amount to read
        stderr = proc.stderr.readline()
        if stderr == '' and stderr == '' and proc.poll() is not None:
            break
        ...do stuff with stdout/stderr here...
```

**Comment by Chris7** [comment]
I spoke too soon. I see you found this thread as well:

http://stackoverflow.com/questions/375427/non-blocking-read-on-a-subprocess-pipe-in-python

:)

**Comment by mfitzp** [comment]
Ha yes that exact one. I had initially implemented it as you say, but if there is nothing in stderr it will block and hang til the script is complete. One other option is postponing stderr checking til the script is complete (though this will stop it being a real "console" log) or piping stdout and stderr together so they can be handled on a single check. Though this will lose use the stdout/stderr split in the DB (and the error checking/display).

It's annoying to have to use threads (especially since this is 2 extra threads per script) so if there is a way around it I would prefer to avoid it.

**Comment by mfitzp** [comment]
So far this live updates the output from stdout/stderr and updates the job status. Additionally, once the job completes the file outputs are also displayed (not currently live updating but will).

@Chris7 I'm seeing occasional errors on the sqlite db lock due to accessing the database from two processes. Updating the stdout/stderr makes the bug more obvious, but it would actually be there regardless due to the need to get the job info into the task execution. It's possible to put a timeout on sqlite access to make it wait. But I'm wondering if there is a 'better' way to do this for the future - we're not really utilising the message queue at all in the current implementation.

**Comment by Chris7** [comment]
@mfitzp sqlite is fundamentally bad at handling multiple users/processes. It's not a bug but a limitation of sqlite.

For the message queue -- we don't know the user will have that as well. But in all honestly, they probably will because things will be painfully slow without celery. Also -- if they are celery, let's hope they are not using sqlite still because the above bug will happen and jobs will be inexplicitly broken.

In the case they do have something like redis/etc. setup, I think that is the place this data should be stored instead of being written to the database. Or at least it should aggregate there with timed writes to the database. A script that just spits out content will be causing thousands of single line writes that need to be possibly sent across the pipe.

I think we should implement it like this:

Provide a setting that defines where this data should be stored, it defaults to the database but has the options of a dictionary of the form:
1) {'Database': dbname}
2) {'CACHE': cache_name}
3) {'MESSAGE_BROKER': ...}

We can use the CELERY_RESULT_BACKEND to get the MESSAGE_BROKER for 3 as well as the django.caches for 2.

**Comment by nzjrs** [comment]
Any news?

**Comment by Chris7** [comment]
I don't know what @mfitzp is up to with this. Currently, it breaks the database transactions with writing so it needs to be refactored to rely on the cache to store information during job updates (a disk based system like redis to avoid crashing), or better multi-thread behavior needs to be incorporated for SQLite databases.

**Comment by mfitzp** [comment]
@nzjrs @Chris7 It is still on my todo list — unfortunately it's on there with a lot of other things. I think finding an alternative route to pass the data and then scheduling writes to the db after some delay (for persistence) is the way to go.

**Comment by mmrasheed** [comment]
I am very much looking forward to the live update feature. I hope you guys make it work soon. :)

**Comment by Chris7** [comment]
@mfitzp I put in the cache work, I think we're good to merge these guys together and in.

**Pull request merged**